### PR TITLE
Improve `JacobianFree` wrapper - new version

### PR DESCRIPTION
**Pull request by vovannikov**
A new version of #590. The initial PR became inapropriate after #378 has been merged due to pre/post.

### CODE REVIEW

File: applications/sintering/include/pf-applications/sintering/residual_wrapper.h [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    {
+      if constexpr (has_evaluate_nonlinear_residual_with_pre_post<
+                      OperatorType,
+                      BlockVectorType>)
+        nonlinear_operator
+          .template evaluate_nonlinear_residual<with_time_derivative>(
+            dst, src, pre_operation, post_operation);
+      else
+        nonlinear_operator
+          .template evaluate_nonlinear_residual<with_time_derivative>(dst, src);
```
Comment (peterrum) [inline applications/sintering/include/pf-applications/sintering/residual_wrapper.h:110]: I think you need to perform pre/post here.

Comment (vovannikov) [reply applications/sintering/include/pf-applications/sintering/residual_wrapper.h:110]: Do you mean that we need to execute `pre_operation()` and `post_operation()` manually via separate `MyMatrixFreeTools::cell_loop_wrapper()` calls with some dummy `cell_operation` if the provided `nonlinear_operator` does not support pre/post? Of course in the case `pre_operation()` and/or `post_operation()` are not empty.

Comment (peterrum) [reply applications/sintering/include/pf-applications/sintering/residual_wrapper.h:110]: If `pre_operation` or `post_operation` are non-empty, they need to be executed. For now, we can assert that they are empty?

Comment (vovannikov) [reply applications/sintering/include/pf-applications/sintering/residual_wrapper.h]: Yea, I see the point. What about 8691efb0a8d3d2ab56dda72f5dc4fa704185dfe5 accompanied by f9d8b7bd3febd45a3d38067eb18ab972edc2199e?

**Pull request merged**